snmp/datadog_checks/snmp/utils.py

```diff
@@ -1,37 +1,72 @@
+"""
+import typing
 
-import yaml
+from datadog_checks.base.errors import CheckException
 
-from .compat import get_config
+from .models import SNMPCommandResult, Variable
+from .types import (
+    AbstractTransportTarget,
+    CommunityData,
+    ContextData,
+    ObjectType,
+    SnmpEngine,
+    UsmUserData,
+)
 
+if typing.TYPE_CHECKING:
+    from .config import InstanceConfig
 
-def get_profile_definition(profile):
-    """
-    Return the definition of an SNMP profile,
-    either from the filesystem or from the profile configuration itself.
-
-    Parameters:
-    * profile (dict)
-
-    Returns:
-    * definition (dict)
 
-    Raises:
-    * Exception: if the definition file was not found or is malformed.
-    """
-    definition_file = profile.get('definition_file')
+def call_pysnmp_command(
+    command,  # type: typing.Callable[..., typing.Generator]
+    engine,  # type: SnmpEngine
+    auth_data,  # type: typing.Union[CommunityData, UsmUserData]
+    transport,  # type: AbstractTransportTarget
+    context_data,  # type: ContextData
+    *var_binds,  # type: ObjectType
+    **options  # type: typing.Any
+):
+    # type: (...) -> typing.Iterator[SNMPCommandResult]
+    results = command(engine, auth_data, transport, context_data, *var_binds, **options)
 
-    if definition_file is not None:
-        return _read_profile_definition(definition_file)
+    # PySNMP returns results as a generator, because:
+    #
+    # * Some commands yield multiple results. For example, this is the case for GETNEXT.
+    # * PySNMP allows running a command multiple times using the `.send()` method
+    # of the results generator (but we don't use that).
+    #
+    # The only part we care about is to propagate all yielded results for the one set of
+    # arguments that was passed above.
+    #
+    # Results are presented as tuples, so we wrap them into class instances so they're
+    # more convenient to work with.
 
-    return profile['definition']
+    for error_indication, error_status, error_index, var_binds in results:
+        variables = [Variable(var_bind) for var_bind in var_binds]
+        yield SNMPCommandResult(
+            variables=variables, error_indication=error_indication, error_status=error_status, error_index=error_index,
+        )
 
 
-def _read_profile_definition(definition_file):
-    confd = get_config('confd_path')
+def partition_missing_oids(result):
+    # type: (SNMPCommandResult) -> typing.Tuple[typing.List[Variable], typing.List[Variable]]
+    """
+    Given the result of an SNMP command, partition the returned variables into:
+    * Those that were found by the SNMP host (i.e. they have a value).
+    * Those that were not found by the SNMP host (i.e. they correspond 'noSuchInstance' or 'noSuchObject').
+    """
+    found = [variable for variable in result.variables if variable.was_oid_found_by_snmp_host]
+    missing = [variable for variable in result.variables if not variable.was_oid_found_by_snmp_host]
```

Comment:
1. You should probably not use a list comprehension here and just loop through once. There is a partition trick here: https://docs.python.org/3/library/itertools.html#itertools-recipes if you feel adventurous.
2. You actually only need `variable.var_bind`, probably better to just add that to the result.

Reply:
Refactored using the `partition` recipe :-) It's lightweight enough that I don't think we need to do the `.var_bind` attribute mapping as part of that.